crates/jiter/src/py_error.rs

```diff
@@ -0,0 +1,210 @@
+    }
+
+    fn __repr__(&self) -> String {
+        format!("JsonParseError({:?})", self.__str__())
+    }
+}
+
+pub(crate) trait MaybeBuildArrayPath: MaybeBuildPath {
+    fn incr_index(&mut self);
+    fn set_index_path(&self, err: PythonJsonError) -> PythonJsonError;
+}
+
+pub(crate) trait MaybeBuildObjectPath: MaybeBuildPath {
+    fn set_key(&mut self, key: &str);
+
+    fn set_key_path(&self, err: PythonJsonError) -> PythonJsonError;
+}
+
+pub(crate) trait MaybeBuildPath {
+    type A: MaybeBuildArrayPath;
+    type O: MaybeBuildObjectPath;
+    fn new_array() -> Self::A;
+    fn new_object() -> Self::O;
+}
+
+pub(crate) struct NoopBuildPath;
+
+impl MaybeBuildPath for NoopBuildPath {
+    type A = NoopBuildPath;
+    type O = NoopBuildPath;
+    fn new_array() -> Self::A {
+        NoopBuildPath
+    }
+
+    fn new_object() -> Self::O {
+        NoopBuildPath
+    }
+}
+
+impl MaybeBuildArrayPath for NoopBuildPath {
+    fn incr_index(&mut self) {}
+
+    fn set_index_path(&self, err: PythonJsonError) -> PythonJsonError {
+        err
+    }
+}
+
+impl MaybeBuildObjectPath for NoopBuildPath {
+    fn set_key(&mut self, _: &str) {}
+
+    fn set_key_path(&self, err: PythonJsonError) -> PythonJsonError {
+        err
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct ActiveBuildPath {
+    index: usize,
+}
+
+impl MaybeBuildPath for ActiveBuildPath {
+    type A = ActiveBuildPath;
+    type O = ActiveObjectBuildPath;
+    fn new_array() -> Self::A {
+        ActiveBuildPath::default()
+    }
+
+    fn new_object() -> Self::O {
+        ActiveObjectBuildPath::default()
+    }
+}
+
+impl MaybeBuildArrayPath for ActiveBuildPath {
+    fn incr_index(&mut self) {
+        self.index += 1;
+    }
+
+    fn set_index_path(&self, mut err: PythonJsonError) -> PythonJsonError {
+        err.add(PathItem::Index(self.index));
+        err
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct ActiveObjectBuildPath {
+    key: String,
+}
+
+impl MaybeBuildPath for ActiveObjectBuildPath {
+    type A = ActiveBuildPath;
+    type O = ActiveObjectBuildPath;
+    fn new_array() -> Self::A {
+        ActiveBuildPath::default()
+    }
+
+    fn new_object() -> Self::O {
+        ActiveObjectBuildPath::default()
+    }
+}
+
+impl MaybeBuildObjectPath for ActiveObjectBuildPath {
+    fn set_key(&mut self, key: &str) {
+        self.key = key.to_string();
+    }
+
+    fn set_key_path(&self, mut err: PythonJsonError) -> PythonJsonError {
+        err.add(PathItem::Key(self.key.clone()));
+        err
+    }
+}
+
+#[derive(Debug, Clone)]
+enum PathItem {
+    Index(usize),
+    Key(String),
+}
+
+impl ToPyObject for PathItem {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        match self {
+            Self::Index(index) => index.to_object(py),
+            Self::Key(str) => str.to_object(py),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct PythonJsonError {
+    pub error: JsonError,
+    path: Option<Vec<PathItem>>,
```

Comment:
Is it worth noting here that the path elements are stored in reverse order? Should we prefer `VecDeque` or `LinkedList` so that we can insert into the front and avoid potential ordering mistakes?